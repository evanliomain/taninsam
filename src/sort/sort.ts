/**
 * @module array=>array
 */
import { ComparaisonResultChoice } from '../@types';

/**
 * Sort an array with the given comparator function.
 * See the [native sort](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort) for more informations.
 * @param comparator The comparator function that defines the sort order. If omitted, the array is sorted according to each character's Unicode code point value, according to the string conversion of each element.
 * @return the function to apply on the array to sort it.
 * @example Without comparator
 * ```
 * sort<number>()([5, 2, 8, 9, 4, 1]) // [1, 2, 4, 5, 8, 9]
 * sort<string>()(['a', 'z', 'e', 'r', 't', 'y']) // ['a', 'e', 'r', 't', 'y', 'z']
 * ```
 * @example With a comparator
 * ```
 * sort<{ a: string, b: number }>()([5, 2, 8, 9, 4, 1]) // [1, 2, 4, 5, 8, 9]
 * ```
 * @example Using the chain
 * ```
 * chain([5, 2, 8, 9, 4, 1])
 *   .chain(sort<number>())
 *   .value() // [1, 2, 4, 5, 8, 9]
 * ```
 */
export function sort<T>(
  comparator?: (d1: T, d2: T) => ComparaisonResultChoice
): (element: ReadonlyArray<T>) => ReadonlyArray<T> {
  return (array: ReadonlyArray<T>) => array.slice().sort(comparator);
}
