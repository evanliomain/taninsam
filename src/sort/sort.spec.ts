import { describe, expect, test } from 'vitest';
import { sort } from './sort';
import { ComparaisonResultChoice } from '../@types/comparaison-result-choice.enum';

describe('sort function', () => {
  const input = [5, 2, 8, 9, 4, 1];

  test('is a pure function', () => {
    sort()(input);
    expect(input).toMatchSnapshot();
  });

  test('[] |> sort === []', () => {
    expect(sort()([])).toMatchSnapshot();
  });

  test('[5, 2, 8, 9, 4, 1] |> sort === [1, 2, 4, 5, 8, 9]', () => {
    expect(sort()(input)).toMatchSnapshot();
  });

  test('["a", "z", "e", "r", "t", "y"] |> sort === ["a", "e", "r", "t", "y", "z"]', () => {
    expect(sort()(['a', 'z', 'e', 'r', 't', 'y'])).toMatchSnapshot();
  });

  test('collection |> sort(comparator)', () => {
    type Sample = { a: number; b: number };
    const comparatorSample = (
      d1: Sample,
      d2: Sample,
    ): ComparaisonResultChoice => {
      if (d1.a === d2.a && d1.b === d2.b) {
        return ComparaisonResultChoice.Equal;
      }
      if (d1.a < d2.a) {
        return ComparaisonResultChoice.Lower;
      }
      if (d2.a < d1.a) {
        return ComparaisonResultChoice.Greater;
      }
      if (d1.a === d2.a) {
        if (d1.b < d2.b) {
          return ComparaisonResultChoice.Lower;
        }
        if (d2.b < d1.b) {
          return ComparaisonResultChoice.Greater;
        }
      }
      return ComparaisonResultChoice.Equal;
    };
    const arraySample: Sample[] = [
      { a: 2, b: 2 },
      { a: 2, b: 1 },
      { a: 3, b: 3 },
      { a: 3, b: 1 },
      { a: 3, b: 2 },
      { a: 1, b: 1 },
      { a: 1, b: 2 },
      { a: 2, b: 3 },
      { a: 1, b: 3 },
    ];
    expect(sort(comparatorSample)(arraySample)).toMatchSnapshot();
  });
});
