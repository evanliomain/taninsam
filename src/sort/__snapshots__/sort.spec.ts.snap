// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort function ["a", "z", "e", "r", "t", "y"] |> sort === ["a", "e", "r", "t", "y", "z"] 1`] = `
Array [
  "a",
  "e",
  "r",
  "t",
  "y",
  "z",
]
`;

exports[`sort function [] |> sort === [] 1`] = `Array []`;

exports[`sort function [5, 2, 8, 9, 4, 1] |> sort === [1, 2, 4, 5, 8, 9] 1`] = `
Array [
  1,
  2,
  4,
  5,
  8,
  9,
]
`;

exports[`sort function collection |> sort(comparator) 1`] = `
Array [
  Object {
    "a": 1,
    "b": 1,
  },
  Object {
    "a": 1,
    "b": 2,
  },
  Object {
    "a": 1,
    "b": 3,
  },
  Object {
    "a": 2,
    "b": 1,
  },
  Object {
    "a": 2,
    "b": 2,
  },
  Object {
    "a": 2,
    "b": 3,
  },
  Object {
    "a": 3,
    "b": 1,
  },
  Object {
    "a": 3,
    "b": 2,
  },
  Object {
    "a": 3,
    "b": 3,
  },
]
`;

exports[`sort function is a pure function 1`] = `
Array [
  5,
  2,
  8,
  9,
  4,
  1,
]
`;
